### PR TITLE
core: avoid `extern type`s in formatting infrastructure

### PR DESCRIPTION
**Pull request by joboet**
@RalfJung [said](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Use.20of.20.60extern.20type.60.20in.20formatting.20machinery/near/446552837):

>How attached are y'all to using `extern type` in the formatting machinery?
Seems like this was introduced a [long time ago](https://github.com/rust-lang/rust/commit/34ef8f5441d5335c4177abd622383ed34a6e9315). However, it's also [not really compatible with Stacked Borrows](https://github.com/rust-lang/unsafe-code-guidelines/issues/256), and only works currently because we effectively treat references-to-extern-type almost like raw pointers in Stacked Borrows -- which of course is unsound, it's not how LLVM works. I was planning to make Miri emit a warning when this happens to avoid cases like [this](https://github.com/rust-lang/rust/issues/126814#issuecomment-2183816373) where people use extern type specifically to silence Miri without realizing what happens. but with the formatting machinery using  extern type, this warning would just show up everywhere...
>
> The "proper" way to do this in Stacked Borrows is to use raw pointers (or `NonNull`).

This PR does just that.

r? @RalfJung

### CODE REVIEW

File: library/core/src/fmt/rt.rs [hunk 1]
```diff
@@ -65,8 +66,11 @@ pub(super) enum Flag {
 }
 
 #[derive(Copy, Clone)]
-enum ArgumentType<'a> {
-    Placeholder { value: &'a Opaque, formatter: fn(&Opaque, &mut Formatter<'_>) -> Result },
+enum ArgumentType {
```
Comment (RalfJung) [inline library/core/src/fmt/rt.rs:69]: Hm, why did you remove the lifetime here?

My plan was to add a `PhantomData<&'a ()>` to represent the borrow. After all, the `value` does have a limited lifetime and that should be reflected in this type, no?

Comment (joboet) [reply library/core/src/fmt/rt.rs:69]: It's added to `Argument` below, I didn't want to add it to the `enum`, that seems unstylish.

File: library/core/src/fmt/rt.rs [hunk 2]
```diff
@@ -162,7 +159,14 @@ impl<'a> Argument<'a> {
...
     pub(super) unsafe fn fmt(&self, f: &mut Formatter<'_>) -> Result {
         match self.ty {
-            ArgumentType::Placeholder { formatter, value } => formatter(value, f),
+            // SAFETY:
+            // `Argument` is constructed so that if `formatter` originally had
+            // the type `fn(&T, ...)` then `value` has type `&T`. Since we use
+            // `value` within the lifetime 'a of the reference and references
+            // and `NonNull` are ABI-compatible, this is completely equivalent
+            // to calling the original function passed to `new` with the original
+            // reference, which is always sound.
```
Comment (RalfJung) [inline library/core/src/fmt/rt.rs:168]: This is actually a data structure invariant, isn't it?
So ideally the reasoning would be: have that invariant documented in the type, and then in `Argument::new` justify why are are establishing the invariant.

Currently this makes an argument about code somewhere else, which is non-local -- that's not great.

**Review by RalfJung — COMMENTED** [review]
Wow, thanks. :) I didn't expect someone else to implement this. :D

File: library/core/src/fmt/rt.rs (hunk 1 above)
Comment (RalfJung) [reply library/core/src/fmt/rt.rs:69]: But this is where the pointer lives. There's a key invariant that `value` is a pointer valid for reading at some dynamic type `T` during lifetime `'a`. By removing the lifetime from ArgumentType you  can't even *state* that invariant as the lifetime is not in scope.

File: library/core/src/fmt/rt.rs (hunk 2 above)
Comment (joboet) [reply library/core/src/fmt/rt.rs]: Fair point, should be fine now.

File: library/core/src/fmt/rt.rs [hunk 3]
```diff
@@ -66,7 +67,13 @@ pub(super) enum Flag {
 
 #[derive(Copy, Clone)]
 enum ArgumentType<'a> {
-    Placeholder { value: &'a Opaque, formatter: fn(&Opaque, &mut Formatter<'_>) -> Result },
+    Placeholder {
+        // INVARIANT: if `formatter` had type `fn(&T, _) -> _` then `value`
+        // was derived from a `&T` with lifetime `'a`.
```
Comment (RalfJung) [inline library/core/src/fmt/rt.rs:72]:
```suggestion
        // INVARIANT: `formatter` has type `fn(&T, _) -> _` for some `T`, and `value`
        // was derived from a `&'a T`.
```

**Comment by RalfJung** [comment]
r=me with the last comment nit fixed
